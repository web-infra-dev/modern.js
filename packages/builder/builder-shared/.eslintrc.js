/** @type {import('eslint').Linter.Config} */
module.exports = {
  extends: ['@modern-js'],
  ignorePatterns: ['compiled/', 'vitest.config.ts'],
  parserOptions: {
    project: require.resolve('./tsconfig.json'),
  },
  rules: {
    'import/order': 0,
    '@typescript-eslint/ban-ts-comment': 0,
    'eslint-comments/no-unlimited-disable': 0,
  },
};
