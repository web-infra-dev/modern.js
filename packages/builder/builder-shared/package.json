{
  "name": "@modern-js/builder-shared",
  "description": "Shared modules of modern.js builder",
  "homepage": "https://modernjs.dev",
  "bugs": "https://github.com/web-infra-dev/modern.js/issues",
  "repository": {
    "type": "git",
    "url": "https://github.com/web-infra-dev/modern.js",
    "directory": "packages/builder/builder-shared"
  },
  "license": "MIT",
  "keywords": [
    "react",
    "framework",
    "modern",
    "modern.js"
  ],
  "engines": {
    "node": ">=14.0.0"
  },
  "version": "2.31.2",
  "jsnext:source": "./src/index.ts",
  "types": "./src/index.ts",
  "main": "./dist/index.js",
  "module": "./dist/index.js",
  "scripts": {
    "prepublishOnly": "only-allow-pnpm",
    "new": "modern new",
    "build": "modern-lib build && node scripts/postCompile.js",
    "dev": "modern-lib build --watch",
    "test": "vitest run",
    "test:watch": "vitest dev --no-coverage",
    "test:ui": "vitest --ui"
  },
  "exports": {
    ".": {
      "jsnext:source": "./src/index.ts",
      "default": "./dist/index.js"
    },
    "./fork-ts-checker-webpack-plugin": {
      "jsnext:source": "./src/exports/fork-ts-checker-webpack-plugin.ts",
      "default": "./dist/exports/fork-ts-checker-webpack-plugin.js"
    },
    "./open": {
      "types": "./compiled/open/index.d.ts",
      "default": "./compiled/open/index.js"
    },
    "./webpack-bundle-analyzer": {
      "types": "./compiled/webpack-bundle-analyzer/index.d.ts",
      "default": "./compiled/webpack-bundle-analyzer/index.js"
    },
    "./babel-loader": "./compiled/babel-loader/index.js",
    "./webpack-5-chain": {
      "types": "./compiled/webpack-5-chain/types/index.d.ts",
      "default": "./compiled/webpack-5-chain/index.js"
    },
    "./pug": {
      "types": "./compiled/pug/index.d.ts",
      "default": "./compiled/pug/index.js"
    },
    "./test-stub": {
      "jsnext:source": "./src/test-stub/index.ts",
      "types": "./dist/test-stub/index.d.ts",
      "default": "./dist/test-stub/index.js"
    },
    "./zod": {
      "jsnext:source": "./src/exports/zod.ts",
      "types": "./dist/exports/zod.d.ts",
      "default": "./dist/exports/zod.js"
    },
    "./css-modules-typescript-loader": {
      "jsnext:source": "./src/loaders/css-modules-typescript-loader.ts",
      "types": "./dist/loaders/css-modules-typescript-loader.d.ts",
      "default": "./dist/loaders/css-modules-typescript-loader.js"
    },
    "./ignore-css-loader": {
      "jsnext:source": "./src/loaders/ignore-css-loader.ts",
      "types": "./dist/loaders/ignore-css-loader.d.ts",
      "default": "./dist/loaders/ignore-css-loader.js"
    }
  },
  "typesVersions": {
    "*": {
      "open": [
        "./compiled/open/index.d.ts"
      ],
      "fork-ts-checker-webpack-plugin": [
        "./dist/exports/fork-ts-checker-webpack-plugin.d.ts"
      ],
      "webpack-bundle-analyzer": [
        "./compiled/webpack-bundle-analyzer/index.d.ts"
      ],
      "webpack-5-chain": [
        "./compiled/webpack-5-chain/types/index.d.ts"
      ],
      "test-stub": [
        "./dist/test-stub/index.d.ts",
        "./src/test-stub/index.ts"
      ],
      "zod": [
        "./dist/exports/zod.d.ts",
        "./src/exports/zod.ts"
      ],
      "pug": [
        "./compiled/pug/index.d.ts"
      ],
      "css-modules-typescript-loader": [
        "./dist/loaders/css-modules-typescript-loader.d.ts",
        "./src/loaders/css-modules-typescript-loader.ts"
      ],
      "ignore-css-loader": [
        "./dist/loaders/ignore-css-loader.d.ts",
        "./src/loaders/ignore-css-loader.ts"
      ]
    }
  },
  "dependencies": {
    "@babel/core": "^7.21.8",
    "@babel/types": "^7.21.5",
    "@babel/parser": "^7.21.8",
    "@modern-js/prod-server": "workspace:*",
    "@modern-js/server": "workspace:*",
    "@modern-js/types": "workspace:*",
    "@modern-js/utils": "workspace:*",
    "@swc/helpers": "0.5.1",
    "fork-ts-checker-webpack-plugin": "8.0.0",
    "acorn": "^8.10.0",
    "caniuse-lite": "^1.0.30001520",
    "cssnano": "6.0.1",
    "htmlparser2": "9.0.0",
    "line-diff": "2.1.1",
    "postcss": "8.4.27",
    "source-map": "^0.7.4",
    "webpack": "^5.88.1",
    "webpack-sources": "^3.2.3",
    "zod": "^3.20.2",
    "zod-validation-error": "1.2.0"
  },
  "devDependencies": {
    "@babel/preset-env": "^7.21.5",
    "@babel/preset-typescript": "^7.21.5",
    "@scripts/vitest-config": "workspace:*",
    "@types/babel__core": "^7.20.0",
    "@types/caniuse-lite": "^1.0.1",
    "@types/node": "^14",
    "html-webpack-plugin": "5.5.3",
    "terser": "5.19.2",
    "typescript": "^5"
  },
  "publishConfig": {
    "registry": "https://registry.npmjs.org/",
    "access": "public",
    "provenance": true,
    "types": "./dist/index.d.ts",
    "exports": {
      ".": "./dist/index.js",
      "./open": "./compiled/open/index.js",
      "./babel-loader": "./compiled/babel-loader/index.js",
      "./webpack-bundle-analyzer": "./compiled/webpack-bundle-analyzer/index.js",
      "./webpack-5-chain": "./compiled/webpack-5-chain/index.js",
      "./test-stub": "./dist/test-stub/index.js",
      "./zod": "./dist/exports/zod.js",
      "./fork-ts-checker-webpack-plugin": "./dist/exports/fork-ts-checker-webpack-plugin.js",
      "./pug": "./compiled/pug/index.js",
      "./css-modules-typescript-loader": "./dist/loaders/css-modules-typescript-loader.js",
      "./ignore-css-loader": "./dist/loaders/ignore-css-loader.js"
    }
  }
}
