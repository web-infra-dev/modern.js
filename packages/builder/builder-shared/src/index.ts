export * from './applyDefaultBuilderOptions';
export * from './constants';
export * from './createHook';
export * from './pluginStore';
export * from './createContext';
export * from './utils';
export * from './fs';
export * from './generateMetaTags';
export * from './getBrowserslist';
export * from './getCssSupport';
export * from './logger';
export * from './mergeBuilderConfig';
export * from './onExitProcess';
export * from './pick';
export * from './regexp';
export * from './types';
export * from './config';
export * from './format';
export * from './devServer';
export * from './prodServer';
export * from './chain';
export * from './schema';
export * from './plugins';
export * from './openBrowser';
export * from './apply';
export * from './fallback';
export * from './getLoaderOptions';
export * from './svgo';
export * from './patch';
export * from './manifest';
export * from './css';
export * from './minimize';
export * from './core-js';
export * from './progress';
export * from './prettyTime';
