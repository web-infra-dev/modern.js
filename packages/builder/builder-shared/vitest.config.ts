import path from 'path';
import { defineConfig } from 'vitest/config';
import { withTestPreset } from '@scripts/vitest-config';

const config = defineConfig({
  test: {
    root: __dirname,
    environment: 'node',
    setupFiles: ['./tests/setup.ts'],
  },
  resolve: {
    alias: {
      '~': path.resolve(__dirname),
      '@': path.resolve(__dirname, 'src'),
    }
  }
});

export default withTestPreset(config);
