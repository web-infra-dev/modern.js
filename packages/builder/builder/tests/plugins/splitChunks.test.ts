import { expect, describe, it, test } from 'vitest';
import * as shared from '@modern-js/builder-shared';
import {
  builderPluginSplitChunks,
  createDependenciesRegExp,
} from '@/plugins/splitChunks';

test('createDependenciesRegExp', () => {
  const cases = {
    'react,react-dom,history':
      /[\\/]node_modules[\\/](react|react-dom|history)[\\/]/,
    '@babel/runtime': /[\\/]node_modules[\\/](@babel\/runtime)[\\/]/,
  };
  for (const [deps, expected] of Object.entries(cases)) {
    const actual = createDependenciesRegExp(...deps.split(','));
    expect(actual).toEqual(expected);
  }
});

describe('plugins/splitChunks', () => {
  const cases = [
    {
      name: 'should set split-by-experience config',
      builderConfig: {
        performance: {
          chunkSplit: {
            strategy: 'split-by-experience',
          },
        },
        output: {
          polyfill: 'entry',
        },
      },
    },
    {
      name: 'should set split-by-experience config correctly when polyfill is off',
      builderConfig: {
        performance: {
          chunkSplit: {
            strategy: 'split-by-experience',
          },
        },
        output: {
          polyfill: 'off',
        },
      },
    },
    {
      name: 'should set split-by-module config',
      builderConfig: {
        performance: {
          chunkSplit: {
            strategy: 'split-by-module',
          },
        },
        output: {
          polyfill: 'entry',
        },
      },
    },
    {
      name: 'should set single-vendor config',
      builderConfig: {
        performance: {
          chunkSplit: {
            strategy: 'single-vendor',
          },
        },
        output: {
          polyfill: 'entry',
        },
      },
    },
    {
      name: 'should set single-size config',
      builderConfig: {
        performance: {
          chunkSplit: {
            strategy: 'split-by-size',
            minSize: 1000,
            maxSize: 5000,
          },
        },
        output: {
          polyfill: 'entry',
        },
      },
    },
    {
      name: 'should set all-in-one config',
      builderConfig: {
        performance: {
          chunkSplit: {
            strategy: 'all-in-one',
          },
        },
        output: {
          polyfill: 'entry',
        },
      },
    },
    {
      name: 'should set custom config',
      builderConfig: {
        performance: {
          chunkSplit: {
            strategy: 'custom',
            forceSplitting: [/react/],
            splitChunks: {
              cacheGroups: {},
            },
          },
        },
        output: {
          polyfill: 'entry',
        },
      },
    },
    {
      name: 'should not split chunks when target is not',
      target: 'node',
    },
  ];

  it.each(cases)('$name', async item => {
    let modifyBundlerChainCb: any;

    const api: any = {
      modifyBundlerChain: (fn: any) => {
        modifyBundlerChainCb = fn;
      },
      getNormalizedConfig: () =>
        item.builderConfig || {
          performance: {
            chunkSplit: {
              strategy: 'split-by-experience',
            },
          },
          output: {
            polyfill: 'entry',
          },
        },
      context: {
        rootPath: __dirname,
      },
    };

    builderPluginSplitChunks().setup(api);

    const chain = await shared.getBundlerChain();

    await modifyBundlerChainCb(chain, {
      target: item.target || 'web',
      isServer: item.target === 'node',
    });

    expect(chain.toConfig()).toMatchSnapshot();
  });
});
