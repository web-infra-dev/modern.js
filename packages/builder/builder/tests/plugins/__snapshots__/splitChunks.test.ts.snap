// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`plugins/splitChunks > 'should allow forceSplitting to be an object' 1`] = `
{
  "optimization": {
    "runtimeChunk": {
      "name": "builder-runtime",
    },
    "splitChunks": {
      "cacheGroups": {
        "axios": {
          "chunks": "all",
          "enforce": true,
          "name": "axios",
          "test": /axios/,
        },
      },
      "chunks": "all",
      "enforceSizeThreshold": 50000,
    },
  },
}
`;

exports[`plugins/splitChunks > 'should not split chunks when target is not' 1`] = `
{
  "optimization": {
    "splitChunks": false,
  },
}
`;

exports[`plugins/splitChunks > 'should set all-in-one config' 1`] = `
{
  "optimization": {
    "splitChunks": false,
  },
}
`;

exports[`plugins/splitChunks > 'should set custom config' 1`] = `
{
  "optimization": {
    "runtimeChunk": {
      "name": "builder-runtime",
    },
    "splitChunks": {
      "cacheGroups": {
        "force-split-0": {
          "chunks": "all",
          "enforce": true,
          "name": "force-split-0",
          "test": /react/,
        },
      },
      "chunks": "all",
      "enforceSizeThreshold": 50000,
    },
  },
}
`;

exports[`plugins/splitChunks > 'should set single-size config' 1`] = `
{
  "optimization": {
    "runtimeChunk": {
      "name": "builder-runtime",
    },
    "splitChunks": {
      "cacheGroups": {},
      "chunks": "all",
      "enforceSizeThreshold": 50000,
      "maxSize": 5000,
      "minSize": 1000,
    },
  },
}
`;

exports[`plugins/splitChunks > 'should set single-vendor config' 1`] = `
{
  "optimization": {
    "runtimeChunk": {
      "name": "builder-runtime",
    },
    "splitChunks": {
      "cacheGroups": {
        "singleVendor": {
          "chunks": "all",
          "enforce": true,
          "name": "vendor",
          "priority": 0,
          "reuseExistingChunk": true,
          "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]/,
        },
      },
      "chunks": "all",
      "enforceSizeThreshold": 50000,
    },
  },
}
`;

exports[`plugins/splitChunks > 'should set split-by-experience config correctly when polyfill is off' 1`] = `
{
  "optimization": {
    "runtimeChunk": {
      "name": "builder-runtime",
    },
    "splitChunks": {
      "cacheGroups": {
        "lib-antd": {
          "name": "lib-antd",
          "priority": 0,
          "reuseExistingChunk": true,
          "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]\\(antd\\)\\[\\\\\\\\/\\]/,
        },
        "lib-arco": {
          "name": "lib-arco",
          "priority": 0,
          "reuseExistingChunk": true,
          "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]\\(@\\?arco-design\\)\\[\\\\\\\\/\\]/,
        },
        "lib-axios": {
          "name": "lib-axios",
          "priority": 0,
          "reuseExistingChunk": true,
          "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]\\(axios\\|axios-\\.\\+\\)\\[\\\\\\\\/\\]/,
        },
        "lib-lodash": {
          "name": "lib-lodash",
          "priority": 0,
          "reuseExistingChunk": true,
          "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]\\(lodash\\|lodash-es\\)\\[\\\\\\\\/\\]/,
        },
        "lib-react": {
          "name": "lib-react",
          "priority": 0,
          "reuseExistingChunk": true,
          "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]\\(react\\|react-dom\\|scheduler\\)\\[\\\\\\\\/\\]/,
        },
        "lib-router": {
          "name": "lib-router",
          "priority": 0,
          "reuseExistingChunk": true,
          "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]\\(react-router\\|react-router-dom\\|@remix-run\\\\/router\\|history\\)\\[\\\\\\\\/\\]/,
        },
      },
      "chunks": "all",
      "enforceSizeThreshold": 50000,
    },
  },
}
`;

exports[`plugins/splitChunks > 'should set split-by-experience config' 1`] = `
{
  "optimization": {
    "runtimeChunk": {
      "name": "builder-runtime",
    },
    "splitChunks": {
      "cacheGroups": {
        "lib-antd": {
          "name": "lib-antd",
          "priority": 0,
          "reuseExistingChunk": true,
          "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]\\(antd\\)\\[\\\\\\\\/\\]/,
        },
        "lib-arco": {
          "name": "lib-arco",
          "priority": 0,
          "reuseExistingChunk": true,
          "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]\\(@\\?arco-design\\)\\[\\\\\\\\/\\]/,
        },
        "lib-axios": {
          "name": "lib-axios",
          "priority": 0,
          "reuseExistingChunk": true,
          "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]\\(axios\\|axios-\\.\\+\\)\\[\\\\\\\\/\\]/,
        },
        "lib-lodash": {
          "name": "lib-lodash",
          "priority": 0,
          "reuseExistingChunk": true,
          "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]\\(lodash\\|lodash-es\\)\\[\\\\\\\\/\\]/,
        },
        "lib-polyfill": {
          "name": "lib-polyfill",
          "priority": 0,
          "reuseExistingChunk": true,
          "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]\\(tslib\\|core-js\\|@babel\\\\/runtime\\|@swc\\\\/helpers\\)\\[\\\\\\\\/\\]/,
        },
        "lib-react": {
          "name": "lib-react",
          "priority": 0,
          "reuseExistingChunk": true,
          "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]\\(react\\|react-dom\\|scheduler\\)\\[\\\\\\\\/\\]/,
        },
        "lib-router": {
          "name": "lib-router",
          "priority": 0,
          "reuseExistingChunk": true,
          "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]\\(react-router\\|react-router-dom\\|@remix-run\\\\/router\\|history\\)\\[\\\\\\\\/\\]/,
        },
      },
      "chunks": "all",
      "enforceSizeThreshold": 50000,
    },
  },
}
`;

exports[`plugins/splitChunks > 'should set split-by-module config' 1`] = `
{
  "optimization": {
    "runtimeChunk": {
      "name": "builder-runtime",
    },
    "splitChunks": {
      "cacheGroups": {
        "vendors": {
          "name": [Function],
          "priority": -10,
          "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]/,
        },
      },
      "chunks": "all",
      "enforceSizeThreshold": 50000,
      "maxInitialRequests": Infinity,
      "minSize": 0,
    },
  },
}
`;
